src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTranslationAction.Actions.cs

```diff
@@ -124,11 +127,76 @@ public ProjectAssemblyNameAction(string assemblyName)
                     _assemblyName = assemblyName;
                 }
 
-                public override Task<Compilation> InvokeAsync(Compilation oldCompilation, CancellationToken cancellationToken)
+                public override Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
                 {
                     return Task.FromResult(oldCompilation.WithAssemblyName(_assemblyName));
                 }
             }
+
+            internal sealed class AddAnalyzerReferencesAction : CompilationTranslationAction
+            {
+                private readonly IEnumerable<AnalyzerReference> _analyzerReferences;
+
+                public AddAnalyzerReferencesAction(IEnumerable<AnalyzerReference> analyzerReferences)
+                {
+                    _analyzerReferences = analyzerReferences;
+                }
+
+                public override TrackedGeneratorDriver TransformGeneratorDriver(TrackedGeneratorDriver generatorDriver)
+                {
+                    var generators = _analyzerReferences.SelectMany(a => a.GetGenerators()).ToImmutableArray();
+                    return new TrackedGeneratorDriver(generatorDriver.GeneratorDriver?.AddGenerators(generators));
```

Comment:
this type is oddly named.  i'm not sure why it's called AddAnalyzerReferencesAction when it's acutaly adding generators to the generator driver.

It also seems weird to have this have to have `?.`   When do we not have a generator driver?  What does this mean in this case, we're just a no-op?  i don't really get it.

Reply:
I was naming the actions for the overall operation that the express (in this case, adding analyzer references) and leaving the implementation detail of what that actually results _in_ internal.

We don't have a generator driver if a previous operation meant we had to throw it out and recompute it (which in the current prototype is quite a lot.) We _could_ have the caller filter this out but then there's no way to null annotate that this is non-null since the type is generally null. I've logged https://github.com/dotnet/roslyn/issues/42815 to track what will ultimately let us get rid of TrackedGeneratorDriver which then means we can make this only get called if we actually have one to update.